### PR TITLE
Inspec autoscalers

### PR DESCRIPTION
**Pull request by slevenick**
Add the autoscaler resource for InSpec. Requires an updated submodule to pull in a rubocop metric change.

This includes some refactoring around plural resources and how they parse properties and nested properties. Parsing should use the same techniques as in singular resources.

Refactoring came up due to a need to reference deeply nested objects for autoscaler tests

-----------------------------------------------------------------
# [all]

### CODE REVIEW

**Comment by slevenick** [comment]
This includes some refactoring around plural resources and how they parse properties and nested properties. Parsing should use the same techniques as in singular resources.

Refactoring came up due to a need to reference deeply nested objects for autoscaler tests

**Comment by rambleraptor** [comment]
Can you break this into one PR for the plural resource changes and one for autoscalers?

File: provider/inspec.rb [hunk 1]
```diff
@@ -280,8 +280,28 @@ def external_attribute(attribute_name, doc_generation = false)
       "'#{YAML.load_file(external_attribute_file)[attribute_name]}'"
     end
 
+    # Replaces Google module name within InSpec resources with GoogleInSpec
+    # to alleviate module namespace conflicts due to dependencies on
+    # Google SDKs
     def inspec_property_type(property)
       property.property_type.sub('Google::', 'GoogleInSpec::')
```
[the remark this answers is not in the document]
Comment (rambleraptor) [reply provider/inspec.rb]: update: ansible is using it too, but it probably should be moved elsewhere? nevermind on that note.

File: templates/inspec/examples/attributes/google_compute_autoscaler.erb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+gcp_project_id = attribute(:gcp_project_id, default: '<%= external_attribute('gcp_project_id') -%>', description: 'The GCP project identifier.')
+gcp_zone = attribute(:gcp_zone, default: '<%= external_attribute('gcp_zone') -%>', description: 'The GCP project zone.')
+instance_group_manager = attribute('instance_group_manager', default: <%= JSON.pretty_generate(grab_attributes['instance_group_manager']) -%>, description: 'Instance group manager definition')
```
Comment (rambleraptor) [inline templates/inspec/examples/attributes/google_compute_autoscaler.erb]: Out of curiosity, what's the difference between calling `pretty_generate` and `.to_json`?

File: templates/inspec/plural_resource.erb [hunk 3]
```diff
@@ -29,7 +29,7 @@ name = "google_#{product_ns.downcase}_#{object.name.underscore}"
   filter_table_config = FilterTable.create
 <% object.all_user_properties.each do |prop| -%>
 
-  <%= "filter_table_config.add(:#{prop.out_name.pluralize}, field: :#{prop.name})" -%>
+  <%= "filter_table_config.add(:#{prop.out_name.pluralize}, field: :#{prop.out_name})" -%>
```
Comment (rambleraptor) [inline templates/inspec/plural_resource.erb]: Why doesn't this break anything?

File: templates/inspec/examples/attributes/google_compute_autoscaler.erb (hunk 2 above)
Comment (slevenick) [reply templates/inspec/examples/attributes/google_compute_autoscaler.erb]: `pretty_generate` adds new lines and spaces and formats the JSON in a (mostly) human readable way. `to_json` just dumps it all on one line

Comment (rambleraptor) [reply templates/inspec/examples/attributes/google_compute_autoscaler.erb]: And we're relying on the fact that JSON.pretty_generate outputs something that happens to look like native Ruby code, right?

(Not a problem with that)

File: templates/inspec/plural_resource.erb (hunk 3 above)
Comment (slevenick) [reply templates/inspec/plural_resource.erb]: The InSpec FilterTable allows users to reference a specific field within a list of hashes. It will grab the value of that field from each hash and return it as a list. This `config.add(name, field)` tells the FilterTable to allow users to reference the field at `field` from each hash by using the key `name`. 

Previously we were just storing the JSON representation of the resources that was retrieved directly from the API, and aliasing the out_name.pluralize to reference the name that the API returns. This worked for top level properties, but not for properties that are nested 2+ levels. 

Using the transformers we are now parsing each resource using the same logic as in singular resources, allowing us to store these using their proper `out_name` as actual ruby objects

Basically this just changes how the internals of plural resources parse and surface attributes, no user facing changes

Comment (rambleraptor) [reply templates/inspec/plural_resource.erb]: Awesome! I was worried that this changed the InSpec control API.

File: templates/inspec/examples/attributes/google_compute_autoscaler.erb (hunk 2 above)
Comment (slevenick) [reply templates/inspec/examples/attributes/google_compute_autoscaler.erb]: I guess we are depending on the fact that Ruby can parse JSON into a hash, regardless of the spacing in that JSON
